lib/AST/GenericSignatureBuilder.cpp

```diff
@@ -4629,8 +4629,35 @@ ConstraintResult GenericSignatureBuilder::addTypeRequirement(
                         ->getDependentType(getGenericParams());
 
       Impl->HadAnyError = true;
-      Diags.diagnose(source.getLoc(), diag::requires_conformance_nonprotocol,
-                     subjectType, constraintType);
+      
+      bool shouldOfferFixit = false;
+      auto DMT = subjectType->getAs<DependentMemberType>();
+      auto GTPT = subjectType->getAs<GenericTypeParamType>();
+
+      if (GTPT) {
+        shouldOfferFixit = isa<ExtensionDecl>(GTPT->getDecl()
+                                                  ->getDeclContext());
+      }
+
+      if (DMT) {
+        shouldOfferFixit = isa<ExtensionDecl>(DMT->getRootGenericParam()
+                                                 ->getDecl()
+                                                 ->getDeclContext());
+      }
+
+      if (shouldOfferFixit) {
+        auto subjectTypeName = DMT ? DMT->getName().str().str() :
+                               subjectType.getString();
+        Diags.diagnose(source.getLoc(), diag::requires_conformance_nonprotocol,
+                       subjectType, constraintType);
+        Diags.diagnose(source.getLoc(),
+                      diag::requires_conformance_nonprotocol_fixit,
+                      subjectTypeName, constraintType.getString())
+             .fixItReplace(source.getLoc(), " == ");
+      } else {
+        Diags.diagnose(source.getLoc(), diag::requires_conformance_nonprotocol,
```

Comment:
This looks the same as line 4651. Can't you unconditionally invoke this before the conditional for the fix-it diagnostic?

Reply:
Thanks, I have fixed it.